### DISCUSSION TITLE
volume fractions similar to STRIPSTAR

### DISCUSSION THREAD

**Discussion by DVangelis** [marcoalopez/GrainSizeTools#17] [Q&A] [answered]
Hi, 
I have a EBSD dataset where I can use to extract the 2D equivalent diameter. Now I want to calculate using Saltykov method, the volume fraction for each grain size class similar to STRIPSTAR software. I believe that the output of stereology.Saltykov method does not return the volume fractions for each calculated grain class. I think what I need is the non-cumulative volume per grain class. How I can extract this data with the GrainSizeTools library? Thanks in advance.

**Comment by marcoalopez** [comment, marked as answer]
Hi, sorry it took me so long to reply. I had missed this. Saltykov's method does not return the fraction per class directly, however, you have the following options:

- 1st option: Use the text_file option, e.g. ``Saltykov(diameters, text_file='foo.csv')``. This will save a table in CSV format in which one of the columns is the accumulated volumes per class. If you subtract the cumulative of the smaller classes for each row, you will get what you are looking for.
- 2nd option: Once you have obtained the midpoints of the classes and the frequencies with the Saltykov function you can do:

```python
mid_points, frequencies = stereology.Saltykov(diameters, return_data=True)

# Calculate the volume for each bin
bin_eq_volume = stereology.diameter_to_volume(mid_points)

# Weight each bin by the volume
vol_weighted_classes = frequencies * bin_eq_volume

# calc volume fraction
vol_fractions = 100 * (vol_weighted_classes / np.sum(bin_eq_volume))

# check that vol_fractions sum 100
np.sum(vol_fractions)
```
It should be noted that this is only an approximation. Once the Saltykov method is applied, it must be assumed that the particle size represented by the midpoint of the class is representative of the particle size population of the whole class, and this may not be strictly true.

_Edit: An alternative option that would not have the problem posed above would be to calculate the volume fractions for specific classes using the two-step method and the function calc_volume_fraction. e.g.:_

```python
# get the volume fraction of a specific grain size
stereology.calc_volume_fraction(lognorm_params=(geomean, std),
                                total_size_range=(min_size, max_size),
                                interest_size_range=(40, 50))
```
_you would need to define the ranges (``interest_size_range``) for each of the calculated classes._

I hope this helps. Let me know if this works for your case.

Cheers.
Marco

**Comment by DVangelis** [comment]
Dear Marco,
Thank you for your reply. After all I came kind of to the same solution by modifying the original stereology.py file.
In the function volume_weighted_cdf inside stereology.py
I also return the vol_weighted_freqs.
It is true that now the midpoint will represent the volume distribution of my sample. And that is highly sensitive to the number of the bins you choose. I tried to find in the literature what is the best number of bins when you apply the Saltykov method, but it seems there is not a specific answer to that, do you have maybe any solution for that ?
From what I understood from the grainsizetools tutorials the two-step method can be used if your dataset can be expressed as a lognormal distribution, otherwise you cannot am I right. When I run the metrics it seems that their is no good fit.
Again thanks for your reply, it was a confirmation that my volume distribution calculations are working.

**Reply by marcoalopez** [reply]
Hi @DVangelis, 

Yes, you are right. Saltykov's method does not provide a method for choosing the optimal number of bins; that is one limitation of the method itself.

There are, however, general methods (i.e. rules of thumb) for choosing an optimal number of classes based on the population. Still, their aim is an optimal representation of the population in a histogram, not Saltykov's method. If you need to limit the number of manual steps and want to use the automatic method for choosing the number of classes, I would try Doane's rule, which does not tend to produce as large a number of classes as other standard methods (the use of many classes tends to destabilise Saltykov's method) and see how it goes.  If you have any further questions, please let me know.

As for the two-step method, you are again right, it requires the distribution to be lognormal.

Best,
Marco
